pkg/tlsconfig/config.go

```diff
@@ -0,0 +1,97 @@
+// Package tlsconfig provides primitives to retrieve secure-enough TLS configurations for both clients and servers.
+//
+// As a reminder from https://golang.org/pkg/crypto/tls/#Config:
+//	A Config structure is used to configure a TLS client or server. After one has been passed to a TLS function it must not be modified.
+//	A Config may be reused; the tls package will also not modify it.
+package tlsconfig
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+)
+
+// NewConfig returns a secure-enough-by-default TLS configuration object.
+func New() *tls.Config {
+	return &tls.Config{
+		// Avoid fallback to SSL protocols < TLS1.0
+		MinVersion:               tls.VersionTLS10,
+		PreferServerCipherSuites: true,
+		CipherSuites: []uint16{
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_RSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+		},
+	}
+}
+
+// certPool returns an X.509 certificate pool from `ca`, the certificate file.
+func certPool(ca string) (*x509.CertPool, error) {
+	// If we should verify the server, we need to load a trusted ca
+	certPool := x509.NewCertPool()
+	file, err := ioutil.ReadFile(ca)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read CA certificate %s: %v", ca, err)
+	}
+	certPool.AppendCertsFromPEM(file)
+	return certPool, nil
+}
+
+// Client returns a TLS configuration meant to be used by a client.
+func Client(enable, verify bool, ca, cert, key string) (*tls.Config, error) {
+	if !enable {
+		return nil, nil
+	}
+	tlsConfig := New()
+	if verify {
+		tlsConfig.InsecureSkipVerify = false
+		certPool, err := certPool(ca)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.RootCAs = certPool
+	} else {
+		tlsConfig.InsecureSkipVerify = true
+	}
+
+	if cert != "" && key != "" {
+		tlsCert, err := tls.LoadX509KeyPair(cert, key)
+		if err != nil {
+			return nil, fmt.Errorf("Could not load X509 key pair: %v. Make sure the key is not encrypted", err)
```

Comment:
@dmcgowan i feel like I found the only case where this is an actual breaking change.
It used to continue if the client certs/keys were not found (or had any kind of error as a matter of fact). But then would error out on the TLS handshake, either because the server has no TLS, or because the server requires client cert. So in those cases, it makes sense to error out early.

However, if the server was configured without CA and thus does not require a Client cert: https://github.com/docker/docker/pull/13636/files#diff-57f93555eddab72218d665bf6f7b239fR88, then failing to read the client certs used to be followed by a successful TLS handshake (without client cert verification), whereas now we would flat out error out.